### ISSUE TITLE
EctoPSQLExtras.query(:calls, Repo) raises

### ISSUE THREAD

**Issue by ZombieHarvester** [pawurb/ecto_psql_extras#8]
Elixir 1.10.4 (compiled with Erlang/OTP 21)
Postgres 12.4
pg_stat_statements installed 1.7

`cache_hit` and `locks` work fine, but some other requests lead to crash. 

```Elixir
iex> EctoPSQLExtras.query(:calls, Repo)
[debug] QUERY OK db=2.7ms queue=1.5ms idle=1490.3ms
/* 10 queries that have the highest frequency of execution */

SELECT query AS qry,
interval '1 millisecond' * total_time AS exec_time,
to_char((total_time/sum(total_time) OVER()) * 100, 'FM90D0') || '%'  AS prop_exec_time,
to_char(calls, 'FM999G999G990') AS ncalls,
interval '1 millisecond' * (blk_read_time + blk_write_time) AS sync_io_time
FROM pg_stat_statements WHERE userid = (SELECT usesysid FROM pg_user WHERE usename = current_user LIMIT 1)
ORDER BY calls DESC
LIMIT 10;
 []
** (ArgumentError) argument error
    (stdlib 3.13.2) :binary.copy(" ", -309)
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:316: TableRex.Renderer.Text.do_render_cell/4
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:288: TableRex.Renderer.Text.render_cell/3
    (elixir 1.10.4) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:277: TableRex.Renderer.Text.render_cell_row/4
    (elixir 1.10.4) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:223: TableRex.Renderer.Text.render_rows/1
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:85: TableRex.Renderer.Text.render/2
    (table_rex 3.0.0) lib/table_rex.ex:26: TableRex.quick_render!/3
    (ecto_psql_extras 0.2.0) lib/ecto_psql_extras.ex:68: EctoPSQLExtras.format/3
```

```Elixir
iex> EctoPSQLExtras.query(:outliers, Repo)
[debug] QUERY OK db=1.0ms queue=0.9ms idle=1996.8ms
/* 10 queries that have longest execution time in aggregate */

SELECT query AS qry,
interval '1 millisecond' * total_time AS exec_time,
to_char((total_time/sum(total_time) OVER()) * 100, 'FM90D0') || '%'  AS prop_exec_time,
to_char(calls, 'FM999G999G999G990') AS ncalls,
interval '1 millisecond' * (blk_read_time + blk_write_time) AS sync_io_time
FROM pg_stat_statements WHERE userid = (SELECT usesysid FROM pg_user WHERE usename = current_user LIMIT 1)
ORDER BY total_time DESC
LIMIT 10;
 []
** (ArgumentError) argument error
    (stdlib 3.13.2) :binary.copy(" ", -20)
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:316: TableRex.Renderer.Text.do_render_cell/4
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:288: TableRex.Renderer.Text.render_cell/3
    (elixir 1.10.4) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:277: TableRex.Renderer.Text.render_cell_row/4
    (elixir 1.10.4) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
    (elixir 1.10.4) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:223: TableRex.Renderer.Text.render_rows/1
    (table_rex 3.0.0) lib/table_rex/renderer/text.ex:85: TableRex.Renderer.Text.render/2
    (table_rex 3.0.0) lib/table_rex.ex:26: TableRex.quick_render!/3
    (ecto_psql_extras 0.2.0) lib/ecto_psql_extras.ex:68: EctoPSQLExtras.format/3
```

Queries work fine, they return results. Seems the problem is in formatting?

**Comment by josevalim** [comment]
What is your ecto_psql_extras version?

**Comment by josevalim** [comment]
This seems to be a bug in table rex somehow? It seems it is trying to do a negative pad. What are the actual query results that you get if you run the query directly?

**Comment by ZombieHarvester** [comment]
@josevalim 

Upgraded to `0.3.0` version

Query from `EctoPSQLExtras.query(:outliers, Repo)` call

```SQL
SELECT query AS query,
interval '1 millisecond' * total_time AS exec_time,
to_char((total_time/sum(total_time) OVER()) * 100, 'FM90D0') || '%'  AS prop_exec_time,
to_char(calls, 'FM999G999G999G990') AS ncalls,
interval '1 millisecond' * (blk_read_time + blk_write_time) AS sync_io_time
FROM pg_stat_statements WHERE userid = (SELECT usesysid FROM pg_user WHERE usename = current_user LIMIT 1)
ORDER BY total_time DESC
LIMIT 10;
```

Returns

![image](https://user-images.githubusercontent.com/923510/96343385-8eb67280-10a2-11eb-8c81-3f61d4ddc01b.png)

**Comment by josevalim** [comment]
I assume it still fails on v0.3.0? In any case we need to convert the percentage to a ratio too so it becomes sortable. I will try to reproduce it at home.

**Comment by ZombieHarvester** [comment]
Yes, this is on 0.3.0

Here is the ecto_sql response, same query as on screenshot above from `EctoPSQLExtras.query(:outliers, Repo)` call:

```Elixir
iex> Ecto.Adapters.SQL.query!(Repo, query)

[debug] QUERY OK db=6.1ms queue=7.7ms idle=1493.3ms

SELECT query AS query,
interval '1 millisecond' * total_time AS exec_time,
to_char((total_time/sum(total_time) OVER()) * 100, 'FM90D0') || '%'  AS prop_exec_time,
to_char(calls, 'FM999G999G999G990') AS ncalls,
interval '1 millisecond' * (blk_read_time + blk_write_time) AS sync_io_time
FROM pg_stat_statements WHERE userid = (SELECT usesysid FROM pg_user WHERE usename = current_user LIMIT 1)
ORDER BY total_time DESC
LIMIT 10; []

%Postgrex.Result{
  columns: ["query", "exec_time", "prop_exec_time", "ncalls", "sync_io_time"],
  command: :select,
  connection_id: 36032,
  messages: [],
  num_rows: 10,
  rows: [
    [
      "SELECT DISTINCT ON ...",
      %Postgrex.Interval{days: 0, microsecs: 915979, months: 0, secs: 1},
      "37.7%",
      "1",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ],
    [
      "with op as ...",
      %Postgrex.Interval{days: 0, microsecs: 657231, months: 0, secs: 1},
      "32.6%",
      "4",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ],
    [
      "UPDATE ...",
      %Postgrex.Interval{days: 0, microsecs: 424749, months: 0, secs: 0},
      "8.3%",
      "5,901",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ],
    [
      "/* Index and table hit rate ...",
      %Postgrex.Interval{days: 0, microsecs: 141735, months: 0, secs: 0},
      "2.8%",
      "3",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ],
    [
      "SELECT name, setting FROM pg_settings WHERE source = $1",
      %Postgrex.Interval{days: 0, microsecs: 123392, months: 0, secs: 0},
      "2.4%",
      "132",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ],
    [
      "UPDATE ...",
      %Postgrex.Interval{days: 0, microsecs: 118463, months: 0, secs: 0},
      "2.3%",
      "5,909",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ],
    [
      "SELECT ...",
      %Postgrex.Interval{days: 0, microsecs: 115025, months: 0, secs: 0},
      "2.3%",
      "3",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ],
    [
      "SELECT ...",
      %Postgrex.Interval{days: 0, microsecs: 108960, months: 0, secs: 0},
      "2.1%",
      "1",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ],
    [
      "INSERT ...",
      %Postgrex.Interval{days: 0, microsecs: 89218, months: 0, secs: 0},
      "1.8%",
      "4",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ],
    [
      "INSERT ...",
      %Postgrex.Interval{days: 0, microsecs: 83403, months: 0, secs: 0},
      "1.6%",
      "58",
      %Postgrex.Interval{days: 0, microsecs: 0, months: 0, secs: 0}
    ]
  ]
}
```

**Comment by josevalim** [comment]
This is an issue with TableRex. Here is how to reproduce it:

```elixir
columns = [:query, :exec_time, :prop_exec_time, :ncalls, :sync_io_time]

rows = [
  [
    "/* 10 queries that have the highest frequency of execution */\n\nSELECT query AS qry,\ninterval $1 * total_time AS exec_time,\nto_char((total_time/sum(total_time) OVER()) * $2, $3) || $4  AS prop_exec_time,\nto_char(calls, $5) AS ncalls,\ninterval $6 * (blk_read_time + blk_write_time) AS sync_io_time\nFROM pg_stat_statements WHERE userid = (SELECT usesysid FROM pg_user WHERE usename = current_user LIMIT $7)\nORDER BY calls DESC\nLIMIT $8",
    "0.001002 seconds",
    "0.012557063201070959",
    "2",
    "0 seconds"
  ],
  [
    "/* 10 queries that have the highest frequency of execution */\n\nSELECT $1 AS query,\ninterval $2 * total_time AS exec_time,\nto_char((total_time/sum(total_time) OVER()) * $3, $4) || $5  AS prop_exec_time,\nto_char(calls, $6) AS ncalls,\ninterval $7 * (blk_read_time + blk_write_time) AS sync_io_time\nFROM pg_stat_statements WHERE userid = (SELECT usesysid FROM pg_user WHERE usename = current_user LIMIT $8)\nORDER BY calls DESC\nLIMIT $9",
    "0.000404 seconds",
    "0.005062856392728202",
    "1",
    "0 seconds"
  ],
  [
    "CREATE EXTENSION pg_stat_statements",
    "0.056623 seconds",
    "0.7098017508490848",
    "1",
    "0 seconds"
  ],
  [
    "SELECT t.oid, t.typname, t.typsend, t.typreceive, t.typoutput, t.typinput,\n       coalesce(d.typelem, t.typelem), coalesce(r.rngsubtype, $1), ARRAY (\n  SELECT a.atttypid\n  FROM pg_attribute AS a\n  WHERE a.attrelid = t.typrelid AND a.attnum > $2 AND NOT a.attisdropped\n  ORDER BY a.attnum\n)\nFROM pg_type AS t\nLEFT JOIN pg_type AS d ON t.typbasetype = d.oid\nLEFT JOIN pg_range AS r ON r.rngtypid = t.oid OR (t.typbasetype <> $3 AND r.rngtypid = t.typbasetype)\nWHERE (t.typrelid = $4)\nAND (t.typelem = $5 OR NOT EXISTS (SELECT $6 FROM pg_catalog.pg_type s WHERE s.typrelid != $7 AND s.oid = t.typelem))",
    "0.020626 seconds",
    "0.2585644066424921",
    "1",
    "0 seconds"
  ],
  [
    "/* 10 queries that have longest execution time in aggregate */\n\nSELECT query AS qry,\ninterval $1 * total_time AS exec_time,\nto_char((total_time/sum(total_time) OVER()) * $2, $3) || $4  AS prop_exec_time,\nto_char(calls, $5) AS ncalls,\ninterval $6 * (blk_read_time + blk_write_time) AS sync_io_time\nFROM pg_stat_statements WHERE userid = (SELECT usesysid FROM pg_user WHERE usename = current_user LIMIT $7)\nORDER BY total_time DESC\nLIMIT $8",
    "0.001118 seconds",
    "0.014013922914623996",
    "1",
    "0 seconds"
  ]
]

TableRex.quick_render!(rows, columns)
```

can you please file it upstream? Thanks!

**Comment by pawurb** [comment]
I've submitted an issue https://github.com/djm/table_rex/issues/38

**Comment by pawurb** [comment]
The issue was caused by newline characters in query strings. It is fixed in 0.3.4.
